dataflow/src/main/java/org/checkerframework/dataflow/cfg/CFGBuilder.java

```diff
@@ -3616,7 +3669,11 @@ public Node visitErroneous(ErroneousTree tree, Void p) {
 
         @Override
         public Node visitExpressionStatement(ExpressionStatementTree tree, Void p) {
-            return scan(tree.getExpression(), p);
+            Node node = scan(tree.getExpression(), p);
```

Comment:
Thanks for exploring this option.
Can you investigate the following alternative: Here, create a local variable and assign non-void expressions to that local variable.
I think that should also cause merging of the two branches and is far less invasive than adding `expressionStatementRootNodes`, for which I don't really see a good use case.